### ISSUE TITLE
better way to deal with history

### ISSUE THREAD

**Issue by samuelcolvin** [samuelcolvin/pydantic#665]
I've just been through about 8 PRs and asked everyone to rebase which is obviously very annoying for them.

Therefore I think we should move to [towncrier](https://pypi.org/project/towncrier/) for history generation.

When creating a PR developers would add a file with details of their change which towncrier (or frankly a dedicated script, it's not that hard) compiles into HISTORY.rst.

The only down side I can think of is that we could not longer release without sitting at a dev machine. Perhaps on release travis could update history "push back" to github with the change?

**Comment by tiangolo** [comment]
I agree it's a good idea to update the workflow with respect to History.

I haven't used towncrier or anything similar, so I can't give much input on that. Up to now, after merging a PR (in FastAPI), I add the release notes (equivalent to History) in a new commit *independent* of the original PR, right after merging it. It's a bit more work than just merging, but hasn't been problematic.

The extra benefit is that a PR wouldn't require to be updated after submitted. Currently, it's necessary to add an extra commit with the PR number, that is only known after submitting.

...Just some extra thoughts on that.

**Comment by dmontagu** [comment]
I hadn't heard of towncrier prior to this, and from the docs alone, I couldn't get a good sense of what it would look like to use it in practice.

After a small amount of digging, I discovered it is used by the twisted project (looks like it was specifically created for twisted); you can see the mechanics "in the wild" by checking out some of their pull requests: https://github.com/twisted/twisted/pulls

They also have a more detailed description of how they expect newsfiles to be incorporated into pull requests here: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles
Note they specifically state it is used to resolve the merge conflicts issue:

> If we just let each author add to the ​NEWS file on every commit, though, we would run into lots of spurious conflicts. To avoid this, we use ​towncrier to manage separate news fragments for each change. 

It actually seems like it would be a nice addition, given:
1) it was appropriately incorporated into the builds/scripts, and
2) there was a clear description of how to comply in the contributing guidelines (it's pretty self-evident how to update the HISTORY.rst, so I guess this is currently unnecessary)

**Comment by dmontagu** [comment]
For what it's worth, you can see towncrier in use "in the wild" by looking at pull requests for twisted: https://github.com/twisted/twisted/pulls

They describe their workflow in more depth here: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

It seems like it would be pretty nice to me, given incorporation into the builds and a little note in the contributing guidelines.

Perhaps the issue of appending the PR number and author to the end of the newsfile content could be automated at merge time?
